pahfit/model.py

```diff
@@ -734,158 +758,169 @@ def tabulate(
+
+        Returns
+        -------
+        Fitter
 
         """
-        if len(self.features) < 2:
-            # Plotting and tabulating works fine, but the code below
-            # will not work with only one component. This can be
-            # addressed later, when the internal API is made agnostic of
-            # the fitting backend (astropy vs our own).
-            raise PAHFITModelError("Fit with fewer than 2 components not allowed!")
-
-        # Some translation rules between astropy model components and
-        # feature table names and values.
-
-        # these have the same value but different (fixed) names
-        param_name_equivalent = {
-            "temperature": "temperature",
-            "fwhm": "fwhm",
-            "x_0": "wavelength",
-            "mean": "wavelength",
-            "tau_sil": "tau",
-        }
-
-        def param_conversion(features_kind, param_name, param_value):
-            # default conversion
-            if param_name in param_name_equivalent:
-                new_name = param_name_equivalent[param_name]
-                new_value = param_value
-            # certain types of features use tau instead of amplitude
-            elif param_name == "amplitude":
-                if features_kind in ["starlight", "dust_continuum", "absorption"]:
-                    new_name = "tau"
+        # Fitting implementation can be changed by choosing another
+        # Fitter class. TODO: make this configurable.
+        fitter = APFitter()
+
+        excluded = self._excluded_features(instrumentname, redshift, x)
+
+        def array3(features_tuple3):
+            return np.array([x for x in features_tuple3])
+
+        for row in self.features[~excluded]:
+            kind = row["kind"]
+            name = row["name"]
+
+            if kind == "starlight":
+                fitter.register_starlight(name, row["temperature"], row["tau"])
+
+            elif kind == "dust_continuum":
+                fitter.register_dust_continuum(name, row["temperature"], row["tau"])
+
+            elif kind == "line":
+                if use_instrument_fwhm:
+                    # one caveat here: redshift. Correct way to do it:
+                    # 1. shift to observed wav; 2. evaluate fwhm at
+                    # oberved wav; 3. shift back to rest frame wav
+                    # (width in rest frame will be narrower than
+                    # observed width)
+                    w_obs = row["wavelength"]["val"] * (1.0 + redshift)
+                    # returned value is tuple (value, min, max). And
+                    # min/max are already masked in case of fixed value
+                    # (output of instrument.resolution is designed to be
+                    # very similar to an entry in the features table)
+                    fwhm = instrument.fwhm(instrumentname, w_obs, as_bounded=True)[
+                        0
+                    ] / (1.0 + redshift)
```

Comment:
You should NOT be de-redshifting here, I think.  We basically look up the FWHM as appropriate at the observed wavelength and use that in the rest frame fit.

Reply:
Note that to do the rest frame fit, we are using modified data, as we correct for the redshift before the fit. As the wavelengths are shortened, the line profiles as seen in the data will become narrower. This change should be reflected in our model, by making the fwhm narrower by the same factor.

Reply:
Good point.  Redshift is annoying.  The model itself needs to shift to the rest frame, and for instrument details, your narrow line got narrower (fractional FWHM is a better parameter for that reason, since it doesn't change with redshift, but that ship has sailed).

For dust features, we specify their positions and widths _in the rest-frame_, so we simply evaluate them there with those (our de-redshifting has de-expanded them).  

When we expand Drude profile widths ala #288, we'll need to do something similar with the teleported line profile width, i.e. providing "the width of the line profile at the observed center of the feature, cast into the rest frame".  We should probably clearly spell this out in a comment here.

Reply:
This has been rephrased as "we do the necessary adjustment as follows"